### ISSUE TITLE
Undefined method <=> for :type:Symbol on display -p --detail

### ISSUE THREAD

**Issue by ghost** [puppetlabs/r10k#310]
Hi

I am trying to display the detailed information about environments and modules using the following command `r10k deploy display -p --detail` and I am getting the following error:

```
- production
        modules:
            - puppetlabs/ntp
[R10K::Action::Deploy::Display - ERROR] undefined method `<=>' for :type:Symbol
            - puppetlabs/stdlib
[R10K::Action::Deploy::Display - ERROR] undefined method `<=>' for :type:Symbol
            - puppetlabs/firewall
[R10K::Action::Deploy::Display - ERROR] undefined method `<=>' for :type:Symbol
            - puppetlabs/concat
[R10K::Action::Deploy::Display - ERROR] undefined method `<=>' for :type:Symbol
            - puppetlabs/apache
[R10K::Action::Deploy::Display - ERROR] undefined method `<=>' for :type:Symbol
            - puppetlabs/ruby
[R10K::Action::Deploy::Display - ERROR] undefined method `<=>' for :type:Symbol
```

My Puppetfile is:

```
forge 'forge.puppetlabs.com'

mod 'puppetlabs/ntp', '3.0.3'
mod 'puppetlabs/stdlib'
mod 'puppetlabs/firewall', '1.4.0'
mod 'puppetlabs/concat', '1.0.0'
mod 'puppetlabs/apache',
  :git => 'https://github.com/puppetlabs/puppetlabs-apache.git',
  :tag => '1.1.0'

mod 'puppetlabs/ruby',
    :git => 'https://github.com/puppetlabs/puppetlabs-ruby.git',
    :tag => '0.4.0'
```

The `r10k deploy environment -p` is working correctly and `r10k deploy display -p` also returns the expected output so I'm not sure if it is a problem with my Puppetfile or an r10k issue

**Comment by adrienthebo** [comment]
Could you run the command with `--trace`?

**Comment by ghost** [comment]
```
$ r10k deploy display --detail -p --trace
Faraday: you may want to install system_timer for reliable timeouts
base (/etc/puppet/environments/)
    - build
        modules:
            - puppetlabs/ntp
[R10K::Action::Deploy::Display - ERROR] undefined method `<=>' for :type:Symbol
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:75:in `sort'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:75:in `visit_module'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:86:in `call'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:86:in `indent'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:73:in `visit_module'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:86:in `call'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:86:in `indent'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:70:in `visit_module'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/visitor.rb:24:in `send'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/visitor.rb:24:in `visit'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/module/base.rb:80:in `accept'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/puppetfile.rb:94:in `accept'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/puppetfile.rb:93:in `each'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/puppetfile.rb:93:in `accept'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:66:in `visit_puppetfile'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/visitor.rb:24:in `send'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/visitor.rb:24:in `visit'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/puppetfile.rb:92:in `accept'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/environment/base.rb:79:in `accept'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:58:in `visit_environment'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:86:in `call'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:86:in `indent'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:56:in `visit_environment'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:86:in `call'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:86:in `indent'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:53:in `visit_environment'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/visitor.rb:24:in `send'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/visitor.rb:24:in `visit'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/environment/base.rb:78:in `accept'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/source/base.rb:64:in `accept'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/source/base.rb:63:in `each'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/source/base.rb:63:in `accept'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:45:in `visit_source'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/visitor.rb:24:in `send'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/visitor.rb:24:in `visit'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/source/base.rb:62:in `accept'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/deployment.rb:108:in `accept'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/deployment.rb:107:in `each'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/deployment.rb:107:in `accept'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:39:in `visit_deployment'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/visitor.rb:24:in `send'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/visitor.rb:24:in `visit'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/deployment.rb:106:in `accept'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/deploy/display.rb:30:in `call'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/runner.rb:23:in `call'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/lib/r10k/action/cri_runner.rb:70:in `call'
/usr/lib/ruby/gems/1.8/gems/cri-2.6.1/lib/cri/command_dsl.rb:223:in `runner'
/usr/lib/ruby/gems/1.8/gems/cri-2.6.1/lib/cri/command.rb:298:in `call'
/usr/lib/ruby/gems/1.8/gems/cri-2.6.1/lib/cri/command.rb:298:in `run_this'
/usr/lib/ruby/gems/1.8/gems/cri-2.6.1/lib/cri/command.rb:251:in `run'
/usr/lib/ruby/gems/1.8/gems/cri-2.6.1/lib/cri/command.rb:264:in `run'
/usr/lib/ruby/gems/1.8/gems/r10k-1.4.1/bin/r10k:7
/usr/bin/r10k:19:in `load'
/usr/bin/r10k:19
```

I get the same output for each module so only copied one here, let me know if you do want the whole thing

**Comment by ghost** [comment]
my ruby version is ruby 1.8.7 (2013-06-27 patchlevel 374) [x86_64-linux] and this is on CentOS 6.5

**Comment by adrienthebo** [comment]
Ruby 1.8.7 doesn't implement `Symbol#<=>`, so this will need to be worked around on Ruby 1.8.7.

As a general note, Ruby 1.8.7 is EOL, Puppet 4 is dropping support for 1.8.7, more and more of the r10k gem dependencies are dropping support as well, and r10k itself won't be supporting 1.8.7 for that much longer, so consider this a gentle nudge to upgrade to a newer version of Ruby if possible. :)

**Comment by nikdoof** [comment]
Might be worth noting in the README that 1.8.7 works with reduced functionality.

**Comment by adrienthebo** [comment]
Copied this issue to https://tickets.puppetlabs.com/browse/RK-36.

**Comment by adrienthebo** [comment]
Documented impending Ruby 1.8.7 deprecation/removal at https://github.com/puppetlabs/r10k/commit/88c66c2a1c1bb2ae65bc8031e39cc1bffb77c731.

**Comment by adrienthebo** [comment]
Released in 1.4.2.
